### PR TITLE
Add HealsPassengers and change UnloadCargo to take an actor reference (or IEnumerable<Actor>).

### PR DESCRIPTION
**Pull request by phrohdoh**
Makes use of new `Cargo::UnloadSpecific` to unload a specific passenger.
The English **TODO** still needs to be done.

Fulfills `Hospital (hurt inf goes in, exits healed)` from #7874.

Currently heals all actors simultaneously, do we want a queue instead (dependent on input order or damage level)?

Depends on #8052.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 1]
```diff
@@ -56,40 +59,42 @@ IEnumerable<CPos> BlockedExitCells(Actor passenger)
 		public override Activity Tick(Actor self)
 		{
 			cargo.Unloading = false;
-			if (IsCanceled || cargo.IsEmpty(self))
+			if (IsCanceled || !actorsToUnload.Any() || !cargo.Passengers.Intersect(actorsToUnload).Any())
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: I'd suggest `!actorsToUnload.Any(x => cargo.Passangers.Contains(x))` here, mostly because it's shorter (these collections don't really have enough elements to make a difference regarding performance).

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 2]
```diff
@@ -56,40 +59,42 @@ IEnumerable<CPos> BlockedExitCells(Actor passenger)
...
 
 			if (cloak != null && cloak.Info.UncloakOnUnload)
 				cloak.Uncloak();
 
-			var actor = cargo.Peek(self);
-			var spawn = self.CenterPosition;
-
-			var exitSubCell = ChooseExitSubCell(actor);
-			if (exitSubCell == null)
+			foreach (var actor in actorsToUnload)
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: You probably want to split the exiting over separate ticks, with a few ticks in between actors, so the ACP/building doesn't spawn 20 infantry at once (actually RA2 structures did that :| ).
RA2 transports, however, had delay between units exiting them, which made it look better (there is only one door, after all).

P.S.: I suggest you do mostly what the old code did, but instead of popping out of `cargo`, take actors from `actorsToUnload`.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: This *is* what the old code did.
It has just been put inside of a `foreach`.
I considered a delay but that is out of scope for this PR.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 3]
```diff
@@ -36,14 +45,19 @@ public class CargoInfo : ITraitInfo, Requires<IOccupySpaceInfo>
 		[Desc("Cursor to display when unable to unload the passengers.")]
 		public readonly string UnloadBlockedCursor = "deploy-blocked";
 
+		[Desc("Unload passengers in reverse load-order?")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: More like "reverse-load order", isn't it?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Yep, will fix.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: This behaviour is also intentional: we moved to Generals style unloading because the earlier one at a time unloading sucked.

**Comment by Phrohdoh** [comment]
Updated the description for `FirstInLastOutUnloading`.

File: OpenRA.Mods.Common/Traits/HealsPassengers.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+{
+	[Desc("Heals all viable passengers simultaneously then optionally unloads fully-healed actors.",
+		"Example use: Hospital structure.")]
+	public class HealsPassengersInfo : ITraitInfo, Requires<CargoInfo>
+	{
+		[Desc("Ticks in-between healing actors.")]
+		public readonly int HealFrequency = 25;
+
+		// TODO: Better English.
+		[Desc("Value to add to the actor's health each HealFrequency ticks..")]
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/HealsPassengers.cs]: You might want to address this.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 5]
```diff
@@ -23,14 +23,17 @@ public class UnloadCargo : Activity
...
+		public UnloadCargo(Actor self, Actor toUnload)
+			: this(self, new[] { toUnload }) { }
+
+		public UnloadCargo(Actor self, IEnumerable<Actor> actorsToUnload)
 		{
 			this.self = self;
 			cargo = self.Trait<Cargo>();
 			cloak = self.TraitOrDefault<Cloak>();
-			this.unloadAll = unloadAll;
+			this.actorsToUnload = actorsToUnload.Where(a => cargo.Passengers.Contains(a));
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: You already throw in `Cargo`, so either this or the throws should go.

File: OpenRA.Game/Traits/Health.cs [hunk 6]
```diff
@@ -188,6 +188,15 @@ public static DamageState GetDamageState(this Actor self)
 			return (health == null) ? DamageState.Undamaged : health.DamageState;
 		}
 
+		public static bool IsDamaged(this Actor self)
```
Comment (penev92) [inline OpenRA.Game/Traits/Health.cs]: This is introducing pretty much duplicate code/behavior. Currently everything uses the above method (`GetDamageState()`) to check if the actor is damaged. I see no reason why you can't use it as well.

**Comment by penev92** [comment]
Code looks mostly OK, but I still need to test this.
Also I've said before that we have **way more** healing/repairing traits than we should, and that we need to cut down on them.
That makes me a bit of hesitant to approve the last part of this PR.

File: OpenRA.Mods.Common/Traits/HealsPassengers.cs (hunk 4 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/HealsPassengers.cs]: Do you have a suggestion other than doing a todo?

File: OpenRA.Game/Traits/Health.cs (hunk 6 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/Health.cs]: There is no reason for me to call the float math if I don't need to, but I can if you guys find it necessary.

**Comment by Phrohdoh** [comment]
Updated.

**Comment by Phrohdoh** [comment]
Updated. No longer mutates `cargo` collection directly.

**Comment by penev92** [comment]
I put a soldier into an APC and unloaded the APC.
The soldier visually came out, then this happened:

>System.InvalidOperationException was unhandled
  HResult=-2146233079
  Message=Cargo for apc 69 does not contain actor e1 70 but tried to unload it.
  Source=OpenRA.Mods.Common
  StackTrace:
       at OpenRA.Mods.Common.Traits.Cargo.UnloadSpecific(Actor self, Actor toUnload)
       at OpenRA.Mods.Common.Activities.UnloadCargo.Tick(Actor self)

**Comment by Phrohdoh** [comment]
Fixed.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs [hunk 7]
```diff
@@ -56,40 +59,44 @@ IEnumerable<CPos> BlockedExitCells(Actor passenger)
...
 
 			if (cloak != null && cloak.Info.UncloakOnUnload)
 				cloak.Uncloak();
 
-			var actor = cargo.Peek(self);
-			var spawn = self.CenterPosition;
-
-			var exitSubCell = ChooseExitSubCell(actor);
-			if (exitSubCell == null)
+			foreach (var actor in actorsToUnload.ToArray())
```
Comment (penev92) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: A much simpler solution that wouls save you the collection transgormations, and then removals, would be to just count how many actors you have unloaded and at the end check if that matches actorsToUnload.Count

File: OpenRA.Game/Traits/Health.cs (hunk 6 above)
Comment (Mailaender) [inline OpenRA.Game/Traits/Health.cs]: DamageState != Undamaged won't do any floating point math.

**Comment by Mailaender** [comment]
> Also I've said before that we have **way more** healing/repairing traits than we should, and that we need to cut down on them.

We did indeed do this with existing traits in the past. https://github.com/OpenRA/OpenRA/pull/2230

File: OpenRA.Game/Traits/Health.cs (hunk 6 above)
Comment (Mailaender) [inline OpenRA.Game/Traits/Health.cs]: Why are you afraid of querying the DamageState? Both performance and rounding errors should not be a problem.

File: OpenRA.Mods.Common/Traits/HealsPassengers.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Heals all viable passengers simultaneously then optionally unloads fully-healed actors.",
+		"Example use: Hospital structure.")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/HealsPassengers.cs]: Can you instead showcase the example use? Place a hospital in the TS test map and setup the traits as you imagine them.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/HealsPassengers.cs]: Sure, give me a bit of time and I'll throw a test commit together.

File: OpenRA.Mods.Common/Activities/UnloadCargo.cs (hunk 7 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: You do realize I am using the `actor` reference, right?

Comment (penev92) [inline OpenRA.Mods.Common/Activities/UnloadCargo.cs]: And you realize you can count stuff in a `foreach`, right?

File: OpenRA.Mods.Common/Traits/HealsPassengers.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+				return;
+
+			if (--countdownTicks > 0)
+				return;
+
+			countdownTicks = info.HealFrequency;
+
+			foreach (var toHeal in damagedPassengers)
+			{
+				var health = toHeal.Trait<Health>();
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/HealsPassengers.cs]: This will cause two lookups for the `Health` trait per passenger per tick.

It would be better to implement `INotifyPassengerEntered` and `INotifyPassengerExited` to cache all passengers' health traits in a dictionary. That will also allow you to remove the redundant `IsDamaged` health extension.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 10]
```diff
@@ -201,11 +217,32 @@ public string VoicePhraseForOrder(Actor self, Order order)
...
+
+			var a = cargo[cargo.Count - 1];
+			return UnloadSpecific(self, a);
+		}
+
+		public Actor UnloadSpecific(Actor self, Actor toUnload)
+		{
+			var idx = cargo.IndexOf(toUnload);
+			if (idx == -1)
+				throw new InvalidOperationException("Cargo for {0} does not contain actor {1} but tried to unload it."
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: I'm not sure if I like `Cargo` throwing exceptions left and right. Why not just return null and make the caller check the return value? The caller needs a check with your approach already (IsEmpty), so turning that check into a null check (if the caller is even interested in it) and thus simplifying `Cargo` would be the better solution IMO.

File: OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs [hunk 11]
```diff
@@ -34,13 +34,15 @@ public TransportProperties(ScriptContext context, Actor self)
...
-		public Actor UnloadPassenger() { return cargo.Unload(Self); }
+		public Actor UnloadPassenger() { return cargo.UnloadLastEntered(Self); }
 
 		[ScriptActorPropertyActivity]
 		[Desc("Command transport to unload passengers.")]
 		public void UnloadPassengers()
 		{
-			Self.QueueActivity(new UnloadCargo(Self, true));
+			var toUnload = cargo.Passengers.ToList();
+			toUnload.Reverse();
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs]: The fact that you have to do this in external, unrelated code to retain the original behaviour smells a bit.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs]: Actually that is the entire point of this; to unload in whatever order you'd like.
And the order we want here is reversed.

Comment (obrakmann) [inline OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs]: Then the `FirstInLastOutUnloading` flag on `Cargo` would be redundant. Policy should either be handled by the trait itself or by the callers, but not both. I very much prefer the former, with appropriate interfaces (ie. public methods and activities) to allow for flexibility.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs]: You're right, although I am not aware of a way for the Lua API to expose the order of passengers, which will be necessary.

When I get the chance I'll clean this up a bit and work on putting everything in Cargo so callers won't know the difference.

**Comment by penev92** [comment]
At least some parts of this will be needed at some point, so adding the future milestone.
